### ISSUE TITLE
Workaround to allow forking after grpc_init()

### ISSUE THREAD

**Issue by dschnabel** [grpc/grpc#21471]
_Note: This is **NOT** a proper fix but rather a quick&dirty workaround to allow process forking after `grpc_init()` has been called. Hopefully there will be a proper fix soon._

The current problem with GRPC as described in many places (e.g. #17695, #13412, #11814) is that once `grpc_init()` has been called in a process, then this process should not be forked. There is a somewhat partial forking support but only if no connection has been established yet.

In my case I have an extension for PHP which is running under Apache. When Apache runs in prefork mode, it dynamically forks child processes to handle the incoming load. This means I need to be able to initialize a new GRPC instance in forked processes even when the parent process already has an active GRPC connection. Since this use case is not yet supported I made a workaround.

Essentially what this workaround does is to reset the internal GRPC state just enough so that a new and clean GRPC instance can be created in a forked child process.

My code change is based on the 1.24.3 release and can be found [here](https://github.com/librato/grpc/compare/v1.24.3...librato:v1.24.3-fork-support).

There are several functions that are called internally to collect active sockets, pthread_once states, mutexes, and conditional variables:
```
void grpc_add_socket_fd(int fd);
void grpc_add_once_init(pthread_once_t *once);
void grpc_add_mutex(pthread_mutex_t *mutex);
void grpc_remove_mutex(pthread_mutex_t *mutex);
void grpc_add_cond(pthread_cond_t *cond);
void grpc_remove_cond(pthread_cond_t *cond);
```
Then there are new globals which hold a pointer to critical internal GRPC states:
```
void **fork_g_event_engine;
void **fork_g_resolver_registry_state;
...
```
---
From the user perspective, all that is needed is to call the newly added function `grpc_clean_after_fork()` before initializing GRPC. Be sure to call it **ONLY ONCE** per process.

`grpc_clean_after_fork()` closes all inherited sockets, resets mutexes, conditional variables, pthread_once states, and other internal states that are left-overs from the parent process. Afterwards GRPC will initialize properly in every forked child.

I'm just sharing my findings here in case somebody else is in a similar situation and wants to give it a try. However, while this workaround works for me, it might not work in other cases.

**Comment by stanley-cheung** [comment]
Coming from #13412, I assume you have tried to enable the following `php.ini` [settings](https://github.com/grpc/grpc/tree/master/src/php#pcntl_fork-support)?

```
grpc.enable_fork_support = 1
grpc.poll_strategy = epoll1
```

**Comment by dschnabel** [comment]
@stanley-cheung thanks for your suggestion but I'm not using the GRPC PHP extension. I have my own PHP extension that starts a long-running thread under the webserver process (e.g. Apache). This thread maintains a GRPC connection that lives beyond any PHP script execution.

The forking is done by Apache, not by PHP. So unfortunately the `php.ini` settings won't help.

**Comment by ericgribkoff** [comment]
It sounds like you are using PHP but not using gRPC PHP, so I'm not very familiar with your specific use case - but gRPC Python and PHP clients support fork *after* connections have been established and have done so since v1.15 (for Python). See the [release notes](https://github.com/grpc/grpc/releases/tag/v1.15.0) and this [comment](https://github.com/grpc/grpc/issues/15334#issuecomment-449102504) broadly describing the API from core's perspective that allows this to work: the child process after fork is responsible for triggering `grpc_shutdown` in core, and the specifics of how to accomplish that in core differ by language (gRPC Python includes its own fork handlers to make this automatic for users). It sounds roughly similar to the approach you describe in your code, with the benefit that it doesn't require hacking gRPC core to add (much) additional global state beyond a boolean that says if a fork has occurred.

It looks like https://github.com/grpc/grpc/blob/master/doc/fork_support.md is in desperate need of an update, but there are numerous now-closed issues centered around this issue before it was fixed in v1.15: please let me know if you have any other questions about this and I will try to get a proper accounting of the current state of fork support into that doc.

**Comment by dschnabel** [comment]
@ericgribkoff that's correct. I'm using PHP but not gRPC PHP. To be more specific, I have a shared library that communicates with a server via gRPC and that library is used by a PHP extension. The same library is also used by other languages, so it's not really PHP-specific.

I had tried using `grpc_shutdown` in the child process but encountered problems with it, since it seems to expect certain gRPC threads to be running which isn't the case after a fork. Basically these parts had issues and caused a crash:
- [Executor::SetThreading](https://github.com/grpc/grpc/blob/v1.24.3/src/core/lib/iomgr/executor.cc#L155)
- [stop_threads](https://github.com/grpc/grpc/blob/v1.24.3/src/core/lib/iomgr/timer_manager.cc#L314)
- [grpc_timer_manager_shutdown](https://github.com/grpc/grpc/blob/v1.24.3/src/core/lib/iomgr/timer_manager.cc#L337)

I [read](https://github.com/grpc/grpc/pull/14647) that fork handlers won't run if multiple threads are calling into gRPC. I do have two stubs in two threads sharing one channel, so maybe that's a reason too why `grpc_shutdown` doesn't work in my case.

Another point that comes to mind is, what if a fork happens while a thread is calling into gRPC? Mutexes can be in a locked state, so destroying them, like it's done in `grpc_shutdown` will result in [undefined behavior](https://pubs.opengroup.org/onlinepubs/009695399/functions/pthread_mutex_destroy.html).

**Comment by ericgribkoff** [comment]
@dschnabel If you are trying to write a gRPC C++ program that supports fork, you will have to obey the same contract as our gRPC Python and gRPC PHP libraries require: namely, no threads can be calling into gRPC core code when the fork occurs, and all gRPC core owned resources must be closed in the child process after fork before the child re-initializes gRPC. This is non-trivial but handled transparently for users by our PHP and Python layers: if you are attempting to re-create the same functionally in your own C++ code I would strongly suggest instead re-architecting things so that your C++ process does not fork. C++ does not have the same constraints that compel PHP/Python users to fork. If your library is not PHP-specific as you comment, you may want to consider moving it out of the PHP process entirely.

Below I discuss more details about the internal fork support in core, but please bear in mind that this is a complicated topic and is a core API intended for gRPC language implementers, not user applications, to consider, and was very specific to PHP and Python and may in some regards not even be compatible with our C++ wrapping layer.

The approach you mention in your original comment:

> grpc_clean_after_fork() closes all inherited sockets, resets mutexes, conditional variables, pthread_once states, and other internal states that are left-overs from the parent process.

Does not scale in a development sense, as internal state in the gRPC library is (a) vast and (b) constantly changing.

> I had tried using grpc_shutdown in the child process but encountered problems with it, since it seems to expect certain gRPC threads to be running which isn't the case after a fork

You can't just directly call `grpc_shutdown` in the child process, as the core API is rather more complicated than that. `grpc_init`/`grpc_shutdown` are basically increment/decrement functions that only have effect when the count goes from 0 to 1 or from 1 to 0 - the first case triggers gRPC core initialization, and the last cleans everything up *but* is dependent on all references to core-allocated objects, such as channels, having actually been properly released by the user application. gRPC PHP, and Python layers manage this complexity automatically for the user.

> Another point that comes to mind is, what if a fork happens while a thread is calling into gRPC?

Fork + multiple threads do not mix well. We avoid this in core by having pthread_atfork handlers that shutdown all core-created threads before the fork occur. As you point out, if multiple *user* threads are calling into gRPC at the time of fork then our handlers will detect this and not run - there is nothing gRPC core can do to make user threads safe during a fork, with multiple failure conditions potentially arising in gRPC core's own state. But in general it's a bad idea for any application to call fork if it knows it's running multiple threads (there are some caveats to this for Python but the gRPC Python layer handles these itself).

Keep in mind that all of the core fork-specific behaviors required the `GRPC_POSIX_FORK_ALLOW_PTHREAD_ATFORK` to be enabled at compile time, and are controlled further at runtime by the `GRPC_ENABLE_FORK_SUPPORT` environment variable.

**Comment by dschnabel** [comment]
I appreciate the insights you are giving into the gRPC core. These make sense but only allow fork support to a certain extent. Based on your post I think the main constraint right now is that an application needs to know in the parent process when a fork is called. Only then can it prevent calls into gRPC until forking is done and the child process can do a proper cleanup using the mechanism you described.

In my case, however, I have no way of knowing in the parent if a fork is about to happen. I can only indirectly determine this in the child by noticing that the PID has changed. But then it's already too late since the fork has happened. And this is not a matter of re-achitecting my application.

I will try to describe in more detail my use case as I think it's not that uncommon:
* My shared library (let's call it profiling.so) is written in C++ and offers an API to applications to collect/process/send profiling events and metrics.
* during initialization profiling.so creates a gRPC channel to an endpoint and two worker threads that process incoming events and metrics. These worker threads each have their own stub on the gRPC channel.
* One application that uses profiling.so is a PHP extention (let's call it php-prof.so). In `MINIT` it initializes profiling.so so that the channel and worker threads are created once. In `MSHUTDOWN` it destroys the channel and worker threads.
* In `RINIT`/`RSHUTDOWN` (PHP request cycle) php-prof.so creates events and metrics and puts them on a queue which gets processed and sent off by the two worker threads started by profiling.so. This happens asynchronously meaning that the PHP request cycle is not affected by the gRPC calls happening in the worker threads.

So far, so good. No forking is involved up to this point. Forking comes into play next:

* PHP runs in Apache as a webserver module.
* Apache is configured to run in prefork mode. In this mode, the Apache parent process forks and manages Apache child processes depending on the request load that's coming in.
* Once forked, there's only one thread running (the apache main thread) in the child process. The first time `RINIT` is called, php-prof.so detects that the PID has changed and re-initializes profiling.so.
* After re-initialization this child process has a new gRPC channel and again two worker threads.

To summarize my setup:
Apache -> PHP module -> PHP extension (php-prof.so) -> C++ library (profiling.so)

Forking happens in Apache and is completely transparent to PHP. Hence profiling.so cannot prevent calls into gRPC until forking is done.

I hope it becomes clear now that it's not a matter of re-achitecting my application. I need to be able to clean up gRPC given any internal gRPC state as Apache can fork at any given time. The workaround I provided in the original post seems to be the solution to this problem. I know it's hacky but wanted to share it for people who face similar dilemmas as I do.

**Comment by ericgribkoff** [comment]
Thanks for the greater detail about your architecture with Apache -> PHP extension -> C++ library -> gRPC. It is indeed (very) difficult to combine fork and multithreading - so difficult that I struggled to find any recommendations for doing so when working on this support for gRPC Python, versus many many recommendations to always avoid this situation. For gRPC PHP and Python it was decided that it was, unfortunately, untenable to disallow fork for our users due to constraints with the language, and so rather a lot of effort was put into making this work.

If you can't change your library structure to move the C++ library out of the PHP process (which you probably can't - in general, something like gRPC would allow the PHP process to communicate with the C++ "library" running in a separate process, but your `profiling.so` name suggests that latency/etc may preclude this) then you essentially have to reimplement our PHP/Python fork handlers in your C++ library, following the contract with core that your fork handlers will release all core resources in the child post-fork. How feasible this is really depends on how your C++ library is using gRPC: you can see how I did it for gRPC Python in https://github.com/grpc/grpc/pull/16264. It's not trivial, but I would argue that this approach (while functionally ~equivalent to the patch you posted originally) is significantly more maintainable as gRPC development continues to move forward. That said, the API with core is really intended for gRPC language implementers and *not* consumers of the C++ surface layer, so there may be significant obstacles with integrating this into your library.

(If exploring this approach, also keep in mind that core must be compiled with the `GRPC_POSIX_FORK_ALLOW_PTHREAD_ATFORK` macro defined and running with the `GRPC_ENABLE_FORK_SUPPORT` environment variable set.)

> Forking happens in Apache and is completely transparent to PHP. Hence profiling.so cannot prevent calls into gRPC until forking is done.

This, fortunately, should not actually be the case. `profiling.so` must register its own fork handlers, which will allow it to pause calls into gRPC prior to the fork, and properly resume/clean-up post-fork in the parent and child. This is how gRPC Python does things (via Cython) in the PR linked above, and how PHP also handles cleanup.

**Comment by dschnabel** [comment]
> This, fortunately, should not actually be the case. profiling.so must register its own fork handlers, which will allow it to pause calls into gRPC prior to the fork, and properly resume/clean-up post-fork in the parent and child.

I see that in the PR #16264 you used `pthread_atfork()` to register your own fork handlers. I had briefly looked into using that as well but didn't persue it further because I thought it was a dead end. But if that's what you used in python and php, maybe that's the way to go. I will give it a try. Thanks for this great tip @ericgribkoff !

**Comment by dschnabel** [comment]
@ericgribkoff I tried using `pthread_atfork()` and registered my own fork handlers (prepare, parent, child):
- the 'prepare' handler acquires the mutexes needed to prevent calls into gRPC
- the 'parent' handler releases the mutexes
- the 'child' handler releases the mutexes

gRPC is compiled with the `GRPC_POSIX_FORK_ALLOW_PTHREAD_ATFORK` macro and the env variable `GRPC_ENABLE_FORK_SUPPORT` is set. I made sure (by using gdb) that the [internal gRPC fork handlers](https://github.com/grpc/grpc/blob/master/src/core/lib/iomgr/fork_posix.cc#L52-L117) are getting called on fork. The way I see it, these internal handlers should take care of cleaning the gRPC state in the child process, correct? Or do I have to do additional work in my own fork handlers to further clean the gRPC state?

The child process crashes with this stacktrace:
```
#0  0x00007fd3b88801e8 in grpc_core::LockfreeEvent::SetReady (this=0x7fd39c0163d0)
    at src/core/lib/iomgr/lockfree_event.cc:210
#1  0x00007fd3b870ad9d in fd_become_writable (fd=0x7fd39c016330)
    at src/core/lib/iomgr/ev_epollex_linux.cc:801
#2  0x00007fd3b870b364 in pollable_process_events (pollset=0x56407a91a360, pollable_obj=0x56407a91a3d0, 
    drain=false) at src/core/lib/iomgr/ev_epollex_linux.cc:900
#3  0x00007fd3b870bfd9 in pollset_work (pollset=0x56407a91a360, worker_hdl=0x0, deadline=33624)
    at src/core/lib/iomgr/ev_epollex_linux.cc:1122
#4  0x00007fd3b8706111 in pollset_work (pollset=0x56407a91a360, worker=0x0, deadline=33624)
    at src/core/lib/iomgr/ev_posix.cc:327
#5  0x00007fd3b85659db in grpc_pollset_work (pollset=0x56407a91a360, worker=0x0, deadline=33624)
    at src/core/lib/iomgr/pollset.cc:48
#6  0x00007fd3b86d621d in run_poller (arg=0x56407a91a2b0, error=0x0)
    at src/core/ext/filters/client_channel/backup_poller.cc:133
#7  0x00007fd3b8568336 in exec_ctx_run (closure=0x56407a91a2e0, error=0x0)
    at src/core/lib/iomgr/exec_ctx.cc:40
#8  0x00007fd3b8568835 in grpc_core::ExecCtx::Flush (this=0x7fd3b5114e00)
    at src/core/lib/iomgr/exec_ctx.cc:158
#9  0x00007fd3b8704356 in run_some_timers () at src/core/lib/iomgr/timer_manager.cc:134
#10 0x00007fd3b870466d in timer_main_loop () at src/core/lib/iomgr/timer_manager.cc:237
#11 0x00007fd3b87047cf in timer_thread (completed_thread_ptr=0x56407a9186e0)
    at src/core/lib/iomgr/timer_manager.cc:284
#12 0x00007fd3b8678b88 in grpc_core::(anonymous namespace)::ThreadInternalsPosix::<lambda(void*)>::operator()(void *) const (__closure=0x0, v=0x56407a917030) at src/core/lib/gprpp/thd_posix.cc:139
#13 0x00007fd3b8678bcd in grpc_core::(anonymous namespace)::ThreadInternalsPosix::<lambda(void*)>::_FUN(void *) () at src/core/lib/gprpp/thd_posix.cc:144
#14 0x00007fd3c9cef6db in start_thread (arg=0x7fd3b5115700) at pthread_create.c:463
#15 0x00007fd3c9a1888f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
```

**Comment by dschnabel** [comment]
Just wanted to note that we've made great experiences with the original workaround I posted bove. Our application which uses the [patched version](https://github.com/librato/grpc/compare/v1.24.3...librato:v1.24.3-fork-support) of gRPC has been deployed on over 16,000 hosts, many of them running in a forked environment. And we've heard of no issues so far.

Since the suggested method of using fork handlers did not work for me, consider my workaround here as an alternative way of allowing fork after grpc_init().
